### DISCUSSION TITLE
2D field distributions from N2F transformation in cylindrical coordinates

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2323] [Q&A]
Hello friends from Meep,

I am using MEEP to simulate a dipole source which is in close proximity to different micro/nano interfaces. Within the materials that I am using I have gold as a reflective surface in order to redirect the emitted field towards the upper half of the structure. By doing so I need to set the Courant factor to a value of 0.1 in order to avoid having my simulation blow in the middle. 

I have tried using the full 3D simulation in this case. Unfortunately the computation just takes a lot of time (more than a day)

In order to speed up my evaluation process I decided to try a 2D simulation in cylindrical coordinates with the dipole source located at r =0. (By following the tutorial ["Extraction Efficiency of a Light-Emitting Diode (LED)"](https://meep.readthedocs.io/en/latest/Python_Tutorials/Local_Density_of_States/#extraction-efficiency-of-a-light-emitting-diode-led). 

Since this is a 2D simulation which is restricted to the rz plane (with phi=0), I was wondering if it is possible to generate and obtain the field components at other different positions with phi not equal to 0. At the end, I would be interested in having the equivalent of a 2D distribution (as a function of x and y).

For this I tried to use a near to far field transformation but it seems that far field evaluation is restricted to points found only within the same plane in where the simulation takes place (the rz plane)

Is there any way to evaluate for field values outside of this plane? Or better said, is there any way to recover or extract 2D field distributions (Ex, Ey and Ez) from 2 cylindrical 2D simulation?

**Comment by smartalecH** [comment]
So first, rather than use a metal reflector, can you reformulate your simulation to somehow use a metallic boundary condition? This removes the stability/resolution requirement typical of metals.

Second, when using cylindrical coordinates, you can transform you field locations to Cartesian coordinates by using the transformation rule (the [wiki](https://en.m.wikipedia.org/wiki/Cylindrical_coordinate_system) article describes this well). You can then interpolate to form a regular grid. 

Also keep in mind that your 2D results have a third dimensional dependence ($\phi$) that depends on the m-number of your simulation. So yes, you can compute the field at any arbitrary point in 3D space, so long as you manually multiply by this phase factor.

**Reply by kalosu** [reply]
Hi @smartalecH , thanks for your feedback!

Yeah I could also replace the gold layer by a metallic boundary condition but still I would be interested in having a more "realistic" simulation and therefore I want to use the 2D simulation in cylindrical coordinates. 

If I understand correctly this is the procedure that I should follow:
1. Run two simulations with different m values (m=1 and m=-1). In my case I have the point like current source at r=0 and z=0 (or any other finite value of z). From what I have seen in MEEP documentation, in this case I can model a dipole oriented along x by defining only a source with the component Er (in cylindrical coordinates with phi=0, Ex=Er)
2. After obtaining the "fourier modes" from these two simulations I can add them up after multiplying them by the azimuthal phase factor in the following manner:
-E(x,y,z) = $E_{m=1}$ (r,z) *exp(j*$\phi$) + $E_{m=-1}$(r,z)*exp(-j*$\phi$) 

3. The azimuthal phase argument $\phi$ is obtained by evaluating something like: $\arctan (y/x)$

Right now I am trying to use this for a simplified model, I just assume that I have a dipole polarized along x which is embedded in air, open and homogeneous space.

In this case, once I apply the summation from the two modes, is the assumption that the Ex component has a one to one correspondence with Er in the full xy plane?

So better said: Is $E_{x}(x,yz)$ = $E_{r,m=1}$ (r,z) *exp(j*$\phi$) + $E_{r,m=-1}$(r,z)*exp(-j*$\phi$) 

Or do I need to take the additional components $E_{\phi}$ and $E_{z}$ for obtaining the final expansion of Ex? 

Maybe you have some reference in where I can take a better look on how to work around with these expansions.

**Comment by kalosu** [comment]
Hi again friends from MEEP,

I was wondering if there is any reference about the expansion of field components in terms of these "cylindrical modes" that you could share with me.

I am trying to generate the electric field components from the obtained Er, Ep and Ez in my 2D cylindrical simulation but I cannot get to obtain similar results as for what I obtain using a 3D simulation in cartesian coordinates. 

In general I notice a few things: (For reference I used the example shown in ["Extraction Efficiency of a Light-Emitting Diode (LED)"](https://meep.readthedocs.io/en/latest/Python_Tutorials/Local_Density_of_States/#extraction-efficiency-of-a-light-emitting-diode-led). Instead of evaluating for the flux I used the "add_dft_fields" object (Er, Ep and Ez) to evaluate for the electric field components)

-After evaluating the results, I noticed something particular about the results for Ep. Differently from Er and Ez, the magnitude of Ep at r=0 is significantly larger. You can take a look to this in the following plot

![E_p_cyl_m_1_r_only](https://user-images.githubusercontent.com/49510503/204334415-9a61d9d3-74ba-4278-a871-5c0ab4fbfb92.png)

I always observe such a large peak for r=0 regardless of the simulation parameters (larger resolution, composition of the materials, Courant factor, etc). I then wonder if this is somehow right. And if so, what is he explanation behind it? 

-As I indicated above, I would be interested in using the solutions from these 2D cylindrical simulations to generate the electric field components in cartesian coordinates (Ex, Ey and Ez). I am trying to do so by first doing a super position of the different "m-modes" for each cylindrical component as: 
$E_{r}$ = $E_{r,m=-1}$ exp(-j $\phi$) + $E_{r,m=0}$ + $E_{r,m=1}$ exp(j $\phi$)

In above expressions, I obtain $\phi$ from taking the arctan of (y/x) for the set of x and y points over which I want to evaluate my 2D field distribution. 

In this case, for the azimuthal phase factor exp(-jm $\phi$) I observe that at points really close to r=0, I obtain a sharp transition in azimuthal phase from one side of the grid to the other. 

![azimuthal_phase_factor_er](https://user-images.githubusercontent.com/49510503/204337468-b5de835d-44d2-4b1f-8391-1e79d80c383b.png)


Taking then the product of Meep's results with this type of azimuthal phase distribution for m=-1 and m=1 and adding these two components results in the distributions that are shown in the following figure:

![E_r_comp_addition_comp](https://user-images.githubusercontent.com/49510503/204340600-82e9aafa-1072-4435-ae35-22f8cd5d2df5.png)

First you can notice the type of "vortex" type of phase that results from the azimuthal phase factor modulation. Following, the addition of the two terms (m=1 and m=-1) results in a phase and amplitude distribution which does not resemble to the expected results (I obtained the expected results from the cartesian 3D simulation of the same problem): 


![Ex_reference_3D_sim](https://user-images.githubusercontent.com/49510503/204342756-5eaa761c-8859-4698-a34b-ed14f827e000.png)

What could be the reason for the discrepancy between the reference 3D simulation and the approach based on the expansion? Have you ever experienced anything similar to this? Do you have any reference publication in where the use of this 2D cylindrical simulations are used for extracting electric field components from point source elements? 

Again, thanks a lot for your help!

**Comment by oskooi** [comment]
> After evaluating the results, I noticed something particular about the results for Ep. Differently from Er and Ez, the magnitude of Ep at r=0 is significantly larger. I always observe such a large peak for r=0 regardless of the simulation parameters (larger resolution, composition of the materials, Courant factor, etc). I then wonder if this is somehow right. And if so, what is he explanation behind it?

This is a property of Maxwell's equations in cylindrical coordinates: for $m \neq 0$, $E_\phi$ at $r = 0$ is $\infty$ (see [equations](https://github.com/NanoComp/meep/issues/2182#issuecomment-1221076961)). You can verify this: as you increase the grid resolution, $E_\phi(r=0) \rightarrow \infty$.

~~Also, note that $E_z$ is *not* defined at $r=0$ in the Yee cell in cylindrical coordinates.~~

**Reply by kalosu** [reply]
I see. Is then possible to generate the 2D field distributions in x and y from the field components obtained through the cylindrical coordinates simulations? Or is the implementation restricted to the evaluation of quantities such as the radiated power?

**Comment by kalosu** [comment]
I think I have managed to generate the 2D distributions in cartesian coordinates that I needed. 
![ex_amp_phase](https://user-images.githubusercontent.com/49510503/204860809-f000bd24-c6dc-4ef8-8096-998bbc26af5b.png)

Nevertheless, I had to manually change the value of Ep at r=0. I set the value of the first position in the Er output vector to the same value as the second position. In this manner I can get rid of the "singularity" at r=0 (specifically, this high value feature is just so for a single point in the output array).

To me, it kind of makes sense that the field should have some sort of smooth and continuous value at each point in space but manually setting the value of this single pixel is something that I am not completely sure about. 

What do you think about this? Is this manual adjustment of the first pixel value acceptable? (This is so just for Ep)

**Comment by stevengj** [comment]
I suspect there is a bug in extracting the `Ep` field right at the origin, because it should be perfectly well-defined and finite in general.  In particular, there are 3 cases at `r=0`, from symmetry:

* `m=0`: fields are rotation-invariant.  At `r=0`, this means that `Ez` and `Hz` can be nonzero, but all the other field components must be zero at the origin (since they are otherwise not rotation-invariant).
* `m=±1`: fields flip sign when you rotate by π.  At `r=0`, this means that `Ep,Er,Hz,Hr` can be nonzero, but the `Ez,Hz` fields must be zero.
* `|m|>1`: all fields go to zero at `r=0`.

(Note that this also requires special handling in the time-stepping equations, because the curl includes terms like `Ez/r`, and you take the limit as `r⟶0`.)

So, we should look at the near2far calculation right at `r=0`, since it might require some special handling.  In the meantime, approximating it by the fields at r=Δr is perfectly reasonable.

**Comment by stevengj** [comment]
Could you please provide a small self-contained test simulation that demonstrates the weird `Ep` value at `r=0` from near2far for `m=1`?

Make it a small simulation that runs quickly, so that we could potentially use it as a unit test after we fix the problem.
